src/annmas/utils/model.py

```diff
@@ -1,16 +1,31 @@
-from pomegranate.callbacks import History, ModelCheckpoint
-from pomegranate import *
+import math
+import re
 
 import matplotlib.pyplot as plt
-from matplotlib import transforms
-
-import pysam
-import pandas as pd
 import numpy as np
+import pandas as pd
+import pysam
+from matplotlib import transforms
+from pomegranate import *
+from pomegranate.callbacks import History, ModelCheckpoint
 
-import re
-import math
-
+array_element_structure = (
```

Comment:
I feel like this might be overcomplicating the issue.  Please correct me if I'm wrong, but after the read is annotated by the HMM, is it possible to simply iteratively segment from the end of the i-th 10x adapter and the beginning of the (i+1)-th 10x adapter, and verify the structure of the read's MAS-seq adapters are also found in the correct order?  Or am I missing something?

Reply:
Since we're modeling the library in this file I wanted to account for what the entire array would look like in an ideal case so that we can use that information to split later.

The "simple splitting" mode is essentially what you describe but without the order constraint. It uses this `array_element_structure` tuple to track what to split on (by choosing leading / trailing parts from each element in  this `array_element_structure`).  I allowed for a configurable number of elements to use as a delimiter (rather than splitting on a single region, like a `10x_adapter`).

The bounded region mode uses this tuple to define each array element as a template and then attempts to match the template (I allowed for a configurable number of missing regions but no mismatches).  It's effectively more strict than the simple splitting mode.